test/jdk/tools/jlink/basic/AllModulePath.java

```diff
@@ -137,7 +164,120 @@ public void testAddModules() throws Throwable {
     }
 
     /*
-     * check the modules linked in the image
+     * No --module-path with --add-modules ALL-MODULE-PATH is an error.
+     */
+    @Test
+    public void noModulePath() {
+        if (isExplodedJDKImage()) {
+            return;
+        }
+        Path targetPath = HELPER.createNewImageDir("all-mod-path-no-mod-path");
+        List<String> allArgs = List.of("--add-modules", "ALL-MODULE-PATH",
+                                       "--output", targetPath.toString());
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        PrintWriter out = new PrintWriter(baos);
+        ByteArrayOutputStream berrOs = new ByteArrayOutputStream();
+        PrintWriter err = new PrintWriter(berrOs);
+        int rc = JLINK_TOOL.run(out, err, allArgs.toArray(new String[] {}));
+        assertTrue(rc != 0);
+        String actual = new String(baos.toByteArray()).trim();
+        assertEquals(actual, "Error: --module-path option must be specified with --add-modules ALL-MODULE-PATH");
+    }
+
+    /*
+     * --module-path not-exist and --add-modules ALL-MODULE-PATH is an error.
+     */
+    @Test
+    public void modulePathEmpty() {
+        if (isExplodedJDKImage()) {
+            return;
+        }
+        Path targetPath = HELPER.createNewImageDir("all-mod-path-not-existing");
+        String strNotExists = "not-exist";
+        Path notExists = Path.of(strNotExists);
+        if (Files.exists(notExists)) {
+            throw new RuntimeException("Test setup error, path must not exist!");
+        }
+        List<String> allArgs = List.of("--add-modules", "ALL-MODULE-PATH",
+                                       "--module-path", notExists.toString(),
+                                       "--output", targetPath.toString());
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        PrintWriter out = new PrintWriter(baos);
+        ByteArrayOutputStream berrOs = new ByteArrayOutputStream();
+        PrintWriter err = new PrintWriter(berrOs);
+        int rc = JLINK_TOOL.run(out, err, allArgs.toArray(new String[] {}));
+        assertTrue(rc != 0);
+        String actual = new String(baos.toByteArray()).trim();
+        assertTrue(actual.startsWith("Error: No module found in module path"));
+        assertTrue(actual.contains(strNotExists));
+    }
+
+    /*
+     * --add-modules ALL-MODULE-PATH with an existing module path and module
+     * limits applied.
+     */
+    @Test
+    public void modulePathWithLimitMods() throws Exception {
```

Comment:
Is this test case different than `testLimitModules` which does
```
jlink --output image1 --add-modules ALL-MODULE-PATH --module-path mods --limit-modules m1
```

`testAddModules` also tests:

```
jlink --output image2 --add-modules m1,test --add-modules ALL-MODULE-PATH --module-path mods --limit-modules java.base
```

Reply:
We have:
- `testLimitModules()` limits on a module from the module path, `m1`. It expects `m1` and it's dependencies in the image.
- `testAddModules()` limits on a module dependency, but has modules from the module path added explicitly: `m1` and `test`. Therefore, expects `m1` and `test` as well as there deps in the image.
- `modulePathWithLimitMods()` limits on a module dependency, but doesn't include any other module with the `--add-modules` option. Therefore, only expects `jdk.jfr` and its dependency, `java.base` to be included in the image.

They are all slightly different and it's useful to test all 3 combinations.

Reply:
I've tried to clarify the cases with comments.

Reply:
Indeed useful tests.   This uncovers an issue about `--limit-modules`.  Thanks.